File: _engine/java/v0_1_0/Prompt.java
// Copyright 2026 GildenEye
// SPDX-License-Identifier: Apache-2.0

package aifunc._engine.java.v0_1_0;

import aifunc.AIFuncException;

import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/** Renders the prompt template by substituting input field placeholders. */
public final class Prompt {

    private Prompt() {}

    private static final Pattern INPUT_FIELD = Pattern.compile("\\{\\{input\\.([a-zA-Z0-9_]+)\\}\\}");
    private static final Pattern BARE_FIELD  = Pattern.compile("\\{\\{([a-zA-Z0-9_]+)\\}\\}");

    /**
     * Renders the prompt template for the given artifact and input.
     *
     * <p>Substitution rules (applied in order):
     * <ol>
     *   <li>{@code {{input_json}}} → full input serialized as JSON
     *   <li>{@code {{input.fieldName}}} → the string value of {@code input["fieldName"]}
     *   <li>{@code {{fieldName}}} → same, but empty string when not found
     * </ol>
     *
     * <p>If {@code injectOutputSchema} is enabled (default), a JSON-schema instruction
     * is appended to the rendered prompt.
     *
     * @throws AIFuncException if no prompt template is found in the artifact
     */
    public static String render(Types.AIFuncArtifact artifact, Map<String, Object> input) {
        String template = selectPrompt(artifact);

        String inputJson = Json.prettyPrint(input);
        String prompt = template.replace("{{input_json}}", inputJson);

        // {{input.fieldName}}
        Matcher m1 = INPUT_FIELD.matcher(prompt);
        StringBuffer sb1 = new StringBuffer();
        while (m1.find()) {
            String field = m1.group(1);
            Object val   = input.get(field);
            m1.appendReplacement(sb1, Matcher.quoteReplacement(val != null ? String.valueOf(val) : m1.group(0)));
        }
        m1.appendTail(sb1);
        prompt = sb1.toString();

        // {{fieldName}}
        Matcher m2 = BARE_FIELD.matcher(prompt);
        StringBuffer sb2 = new StringBuffer();
        while (m2.find()) {
            String field = m2.group(1);
            Object val   = input.get(field);
            m2.appendReplacement(sb2, Matcher.quoteReplacement(val != null ? String.valueOf(val) : ""));
        }
        m2.appendTail(sb2);
        prompt = sb2.toString();

        if (shouldInjectSchema(artifact)) {
            @SuppressWarnings("unchecked")
            Map<String, Object> outputSchema = (Map<String, Object>) artifact.getApi().get("output");
            prompt = prompt + "\n\n" + buildSchemaInstruction(outputSchema);
        }

        return prompt;
    }

    // -------------------------------------------------------------------------
    // Helpers
    // -------------------------------------------------------------------------

    private static String selectPrompt(Types.AIFuncArtifact artifact) {
        if (artifact.getPrompt() != null && !artifact.getPrompt().isBlank()) {
            return artifact.getPrompt();
        }
        Map<String, String> prompts = artifact.getPrompts();
        if (prompts != null) {
            String general = prompts.get("general");
            if (general != null && !general.isBlank()) return general;
            for (String p : prompts.values()) {
                if (p != null && !p.isBlank()) return p;
            }
        }
        throw new AIFuncException("Artifact missing prompt template");
    }

    @SuppressWarnings("unchecked")
    private static boolean shouldInjectSchema(Types.AIFuncArtifact artifact) {
        Map<String, Object> api = artifact.getApi();
        if (api != null) {
            Object inject = api.get("injectOutputSchema");
            if (inject instanceof Boolean && !((Boolean) inject)) return false;
        }
        Map<String, Object> pkg = artifact.getPackage();
        if (pkg != null) {
            Object opts = pkg.get("engineOptions");
            if (opts instanceof Map) {
                Object inject = ((Map<String, Object>) opts).get("injectOutputSchema");
                if (inject instanceof Boolean && !((Boolean) inject)) return false;
            }
        }
        return true;
    }

    private static String buildSchemaInstruction(Map<String, Object> schema) {
        return "Please respond with a JSON object that matches the following schema:\n\n"
                + Json.prettyPrint(schema)
                + "\n\nYour response must be valid JSON only, with no additional text.";
    }
}

File: _engine/csharp/v0_1_0/Prompt.cs
// Copyright 2026 GildenEye
// SPDX-License-Identifier: Apache-2.0

using System;
using System.Collections;
using System.Collections.Generic;
using System.Text.RegularExpressions;
using Aifunc;

namespace Aifunc.Engine.Csharp.V0_1_0;

/// <summary>Renders the prompt template by substituting input field placeholders.</summary>
public static class Prompt
{
    private static readonly Regex InputField = new(@"\{\{input\.([a-zA-Z0-9_]+)\}\}", RegexOptions.Compiled);
    private static readonly Regex BareField  = new(@"\{\{([a-zA-Z0-9_]+)\}\}", RegexOptions.Compiled);

    /// <summary>
    /// Renders the prompt template for the given artifact and input.
    ///
    /// Substitution rules (applied in order):
    /// <list type="number">
    ///   <item><c>{{input_json}}</c> → full input serialized as JSON</item>
    ///   <item><c>{{input.fieldName}}</c> → the string value of <c>input["fieldName"]</c></item>
    ///   <item><c>{{fieldName}}</c> → same, but empty string when not found</item>
    /// </list>
    ///
    /// If <c>injectOutputSchema</c> is enabled (default), a JSON-schema instruction
    /// is appended to the rendered prompt.
    /// </summary>
    /// <exception cref="AIFuncException">Thrown if no prompt template is found in the artifact.</exception>
    public static string Render(Types.AIFuncArtifact artifact, Dictionary<string, object?> input)
    {
        var template = SelectPrompt(artifact);

        var inputJson = Json.PrettyPrint(input);
        var prompt = template.Replace("{{input_json}}", inputJson);

        // {{input.fieldName}}
        prompt = InputField.Replace(prompt, m =>
        {
            var field = m.Groups[1].Value;
            if (input.TryGetValue(field, out var val) && val != null)
            {
                return Convert.ToString(val) ?? m.Value;
            }
            return m.Value;
        });

        // {{fieldName}}
        prompt = BareField.Replace(prompt, m =>
        {
            var field = m.Groups[1].Value;
            if (input.TryGetValue(field, out var val) && val != null)
            {
                return Convert.ToString(val) ?? "";
            }
            return "";
        });

        if (ShouldInjectSchema(artifact))
        {
            Dictionary<string, object?>? outputSchema = null;
            if (artifact.Api != null && artifact.Api.TryGetValue("output", out var outVal))
            {
                outputSchema = AsDict(outVal);
            }
            prompt = prompt + "\n\n" + BuildSchemaInstruction(outputSchema);
        }

        return prompt;
    }

    private static string SelectPrompt(Types.AIFuncArtifact artifact)
    {
        if (!string.IsNullOrWhiteSpace(artifact.Prompt))
        {
            return artifact.Prompt!;
        }
        var prompts = artifact.Prompts;
        if (prompts != null)
        {
            if (prompts.TryGetValue("general", out var general) && !string.IsNullOrWhiteSpace(general))
            {
                return general;
            }
            foreach (var p in prompts.Values)
            {
                if (!string.IsNullOrWhiteSpace(p)) return p;
            }
        }
        throw new AIFuncException("Artifact missing prompt template");
    }

    private static bool ShouldInjectSchema(Types.AIFuncArtifact artifact)
    {
        var api = artifact.Api;
        if (api != null && api.TryGetValue("injectOutputSchema", out var inject)
            && inject is bool injectBool && !injectBool)
        {
            return false;
        }
        var pkg = artifact.Package;
        if (pkg != null && pkg.TryGetValue("engineOptions", out var opts) && AsDict(opts) is { } optsMap)
        {
            if (optsMap.TryGetValue("injectOutputSchema", out var inject2)
                && inject2 is bool injectBool2 && !injectBool2)
            {
                return false;
            }
        }
        return true;
    }

    private static string BuildSchemaInstruction(Dictionary<string, object?>? schema)
    {
        return "Please respond with a JSON object that matches the following schema:\n\n"
               + Json.PrettyPrint(schema)
               + "\n\nYour response must be valid JSON only, with no additional text.";
    }

    private static Dictionary<string, object?>? AsDict(object? v)
    {
        if (v is Dictionary<string, object?> d) return d;
        if (v is System.Collections.IDictionary raw)
        {
            var converted = new Dictionary<string, object?>();
            foreach (System.Collections.DictionaryEntry e in raw)
            {
                converted[Convert.ToString(e.Key) ?? ""] = e.Value;
            }
            return converted;
        }
        return null;
    }
}
